src/Build/Evaluation/LazyItemEvaluator.cs

```diff
@@ -219,17 +219,299 @@ public void MarkAsReferenced()
+                {
+                    _version = version;
+                    _value = item;
+                }
+
+                public void Add(int version, I item)
+                {
+                    if (_value is List<I> list)
+                    {
+                        if (version != _version)
+                        {
+                            list = new List<I>(list);
+                        }
+                        list.Add(item);
+                    }
+                    else
+                    {
+                        list = new List<I>
+                        {
+                            (I)_value,
+                            item
+                        };
+                    }
+                    _version = version;
+                    _value = list;
+                }
+
+                public IEnumerator<I> GetEnumerator()
+                {
+                    if (_value is I item)
+                    {
+                        yield return item;
+                    }
+                    else if (_value is IEnumerable<I> enumerable)
+                    {
+                        foreach (I enumerableItem in enumerable)
+                        {
+                            yield return enumerableItem;
+                        }
+                    }
+                }
+
+                System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => GetEnumerator();
+            }
+
+            /// <summary>
+            /// A mutable and enumerable version of <see cref="OrderedItemDataCollection"/>.
+            /// </summary>
+            internal sealed class Builder : IEnumerable<ItemData>
+            {
+                /// <summary>
+                /// The current version of the collection.
+                /// </summary>
+                private int _version;
+
+                /// <summary>
+                /// The list of items in the collection. Defines the enumeration order.
+                /// </summary>
+                private ImmutableList<ItemData>.Builder _listBuilder;
+
+                /// <summary>
+                /// A dictionary of items keyed by their normalized value.
+                /// </summary>
+                private ImmutableDictionary<string, DictionaryValue>.Builder _dictionaryBuilder;
+
+                internal Builder(int version, ImmutableList<ItemData>.Builder listBuilder, ImmutableDictionary<string, DictionaryValue>.Builder dictionaryBuilder)
+                {
+                    _version = version;
+                    _listBuilder = listBuilder;
+                    _dictionaryBuilder = dictionaryBuilder;
+                }
+
+                #region IEnumerable implementation
+
+                public IEnumerator<ItemData> GetEnumerator() => _listBuilder.GetEnumerator();
+
+                System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => _listBuilder.GetEnumerator();
```

Comment:
In example, allocations can be avoided here with the ImmutableList enumerator directly:

```suggestion
                ImmutableList<ItemData>.Enumerator GetEnumerator() => _listBuilder.GetEnumerator();
                IEnumerator<ItemData> IEnumerable<T>.GetEnumerator() => _listBuilder.GetEnumerator();

                System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => _listBuilder.GetEnumerator();
```

Reply:
Done, thank you!